mri/reconstructors/calibrationless.py

```diff
@@ -0,0 +1,107 @@
+        have 2 functions, op(x) and adj_op(coeff) which implements the
+        operator and adjoint operator. For wavelets, this can be object of
+        class WaveletN or WaveletUD2 from mri.operators .
+        If None, sym8 wavelet with nb_scale=3 is chosen.
+    regularizer_op: operator, (optional default None)
+        Defines the regularization operator for the regularization function H.
+        If None, the  regularization chosen is Identity and the optimization
+        turns to gradient descent.
+    gradient_formulation: str between 'analysis' or 'synthesis',
+        default 'synthesis'
+        defines the formulation of the image model which defines the gradient.
+    lips_calc_max_iter: int, default 10
+        Defines the maximum number of iterations to calculate the lipchitz
+        constant
+    num_check_lips: int, default 10
+        Number of iterations to check if the lipchitz constant is correct
+    lipschitz_cst: int, default None
+        The user specified lipschitz constant. If this is not specified,
+        it is calculated using PowerMethod
+    n_jobs : int, default 1
+        The number of cores to be used for faster reconstruction
+    verbose: int, default 0
+        Verbosity level.
+            1 => Print basic debug information
+            5 => Print all initialization information
+            20 => Calculate cost at the end of each iteration.
+                NOTE : This is computationally intensive.
+            30 => Print the debug information of operators if defined by class
+    Note:
+    -----
+    The user is expected to specify the either prox_op or mu to obtain
+    reconstructions, else the above equations lose the regularization terms
+    resulting in inverse transform as solution.
+    The reconstruction in this case proceeds with a warning.
+    """
+
+    def __init__(self, fourier_op, linear_op=None, regularizer_op=None,
+                 gradient_formulation="synthesis", lips_calc_max_iter=10,
+                 num_check_lips=10, lipschitz_cst=None, n_jobs=1, verbose=0):
+        if linear_op is None:
+            linear_op = WaveletN(
+                # TODO change nb_scales to max_nb_scale - 1
+                wavelet_name="sym8",
+                nb_scale=3,
+                dim=len(fourier_op.shape),
+                n_coils=fourier_op.n_coils,
+                n_jobs=n_jobs,
+                verbose=bool(verbose >= 30),
+            )
+        # Ensure that we are in right multichannel config
+        if fourier_op.n_coils != linear_op.n_coils:
+            raise ValueError("The value of n_coils for fourier and wavelet "
+                             "operation must be same for "
+                             "calibrationless reconstruction!")
+        if gradient_formulation == 'analysis':
+            grad_class = GradAnalysis
+        elif gradient_formulation == 'synthesis':
+            grad_class = GradSynthesis
```

Comment:
You could avoid all of this if the base reconstructor takes the gradient operator as input

Reply:
No, That was how I had coded it initially. But that just caused way too much redundant code.
We would just end up with a bloated code in every derived class

Reply:
I'm not confortable with this solution at all, I guess that we might have a better option

Reply:
But the other option is to have codes like this:

```python
if gradient_formulation == 'analysis':
    grad_op = GradAnalysis(
            fourier_op=self.fourier_op,
            lips_calc_max_iter=self.lips_calc_max_iter,
            lipschitz_cst=self.lipschitz_cst,
            num_check_lips=self.num_check_lips,
            verbose=self.verbose,
        )
elif gradient_formulation == 'synthesis':
    grad_op = GradSynthesis(
            fourier_op=self.fourier_op,
            lips_calc_max_iter=self.lips_calc_max_iter,
            lipschitz_cst=self.lipschitz_cst,
            num_check_lips=self.num_check_lips,
            verbose=self.verbose,
            linear_op=linear_op,
)
```

This would be present in every reconstructor class!
I am surely not comfortable with this. We can discuss on a better option off hand.

Reply:
@zaccharieramzi and I had reformatted this. @zaccharieramzi your comments? 
Would you prefer redundant code or understandable code?

Reply:
I think this is really nice

Reply:
Hmm, @LElgueddari do you want to discuss further? We could meet once on this? I see that you have accepted the review also. If you are fine with this, can we merge once the tests pass?